File: matlab_src/run_phantom_t2.m
%% run_phantom_t2.m
%  Phantom Experiment: T2 Mapping Validation
%
%  DESCRIPTION:
%  Performs T2 mapping and uncertainty quantification on phantom data.
%  It processes scans with different acceleration factors, compares them against
%  a Gold Standard SESE reference, and saves the resulting maps and plots.
%
%  SCENARIOS:
%  1. Methods: LRT vs Bayesian
%  2. B1 Information: Free fit vs Constrained (Measured B1 +/- 10%)
%  3. Acceleration: 8192 (NUFFT), 384 (LLR), 192 (LLR) views
%
%  OUTPUTS:
%  - /images: T2 Maps and Uncertainty Maps (.png/.tiff)
%  - /plots:  Correlation plots (Reference vs Estimated)

restoredefaultpath;
clear; clc; close all;

%% 1. Setup Paths
[script_dir, ~, ~] = fileparts(mfilename('fullpath'));
addpath(genpath(script_dir));

repo_root = fileparts(script_dir);
data_dir  = fullfile(repo_root, 'data', 'phantom_t2');
dict_dir  = fullfile(repo_root, 'data', 'dictionaries');

%% 2. Experiment Configuration
% Scan Definitions (Filename must match what is in /data/phantom_t2/)
scans(1).name = '8192_NUFFT'; scans(1).file = 'scan_8192_nufft.mat'; dicts(1).file = 't2_phantom_TE.mat'; 
scans(2).name = '384_LLR';    scans(2).file = 'scan_384_llr.mat'; dicts(2).file = 't2_phantom_PC.mat'; 
scans(3).name = '192_LLR';    scans(3).file = 'scan_192_llr.mat'; dicts(3).file = 't2_phantom_PC.mat'; 

% Processing Settings
alpha_lvl = 0.05;        % 95% CI
trunc_factor = 3;        % TE Truncation threshold (3 * T2)
t2_disp_range = [0 150]; % Display range for T2 maps (ms)
uq_disp_range = [0 40];  % Display range for Uncertainty maps (ms)
fov = [96 96];           % cropped fov (voxels)
use_identity_cov = false; % Set to true to force Identity Covariance (Failure mode test)

% B1 Modes to Iterate
b1_modes = {'Free', 'Range'}; 

% Output Directory Logic
if use_identity_cov
    out_dir = fullfile(repo_root, 'matlab_output', 'phantom_t2_results_identity');
    fprintf('Running in IDENTITY COVARIANCE mode. Saving to: %s\n', out_dir);
else
    out_dir = fullfile(repo_root, 'matlab_output', 'phantom_t2_results');
    fprintf('Running in STANDARD COVARIANCE mode. Saving to: %s\n', out_dir);
end

if ~isfolder(out_dir); mkdir(out_dir); end
if ~isfolder(fullfile(out_dir, 'images')); mkdir(fullfile(out_dir, 'images')); end
if ~isfolder(fullfile(out_dir, 'plots')); mkdir(fullfile(out_dir, 'plots')); end

%% 3. Load Common Resources
fprintf('Loading Reference Data and Dictionary...\n');

try
    % Load References: sese_map (T2 Ref), b1_map (Measured B1)
    load(fullfile(data_dir, 'reference_maps.mat'), 'sese_t2map', 'sese_contrast', 'tfl_b1map');
    % Load ROIs: roi_masks (Struct or 3D array of logical masks)
    load(fullfile(data_dir, 'roi_masks.mat'), 'roi_masks');
catch
    error('Missing reference/dictionary files in %s or %s', data_dir, dict_dir);
end

% crop images
sese_t2map = crop_image(sese_t2map, fov); 
sese_contrast = crop_image(sese_contrast, fov); 
tfl_b1map = crop_image(tfl_b1map, fov); 

% Create a display mask (background suppression)
display_mask = sese_contrast(:,:,1) > 0.1*max(sese_contrast(:)); % Simple threshold on reference

% Container for Plotting Statistics
plot_db = struct(); 

%% 4. Main Processing Loop
for s_idx = 1:length(scans)
    scan = scans(s_idx);
    dict = dicts(s_idx); 
    fprintf('\n--- Processing Scan: %s ---\n', scan.name);
    
    % Load Contrast Data
    try
        load(fullfile(data_dir, scan.file), 'contrast', 'header');
    catch
        warning('Skipping %s (File not found)', scan.file); continue;
    end
    % Load Dictionary: D
    try
        load(fullfile(dict_dir, dict.file), 'D');
    catch
        warning('Skipping %s (Dict not found)', dict.file); continue;
    end
    
    % Prepare Data (Scale and Estimate Noise)
    contrast_dbl = crop_image(double(contrast) * 1e0, fov); 
    
    if use_identity_cov
        % Use Identity Matrix (assume i.i.d. noise)
        sigma = eye(size(contrast_dbl, 3));
    else
        % Estimate Covariance Empirically
        sigma = estimateNoiseCovariance(contrast_dbl, 10);
    end
    
    TE_array = (1:header.etl) * header.esp;
    
    for b_idx = 1:length(b1_modes)
        mode = b1_modes{b_idx};
        use_map = strcmp(mode, 'Range');
        fprintf('   > Mode: B1 %s\n', mode);
        
        % --- Configure Options ---
        ops = struct('alpha', alpha_lvl, 'te_truncation', size(contrast, 3) == header.etl, ...
                     'te_array', TE_array, 'trunc_factor', trunc_factor);
        
        if use_map
            % Updated Logic: Use 'range' mode with +/- 10% bounds
            ops.b1_mode = 'range';
            % Create [Nx, Ny, 2] map defining the valid interval per voxel
            ops.b1_input = cat(3, 0.9 * tfl_b1map, 1.1 * tfl_b1map); 
        else
            ops.b1_mode = 'none';
        end
        
        % --- Run Solvers ---
        % 1. LRT
        t0 = tic;
        [lrt_maps, lrt_stats] = fit_mri_params_lrt(contrast_dbl, sigma, D, ops);
        lrt_uq = lrt_stats.q_ci(:,:,2) - lrt_stats.q_ci(:,:,1); % Width of CI
        fprintf('     LRT: %.2fs | ', toc(t0));
        
        % 2. Bayesian
        t0 = tic;
        [bayes_maps, bayes_stats] = fit_mri_params_bayesian(contrast_dbl, sigma, D, ops);
        bayes_uq = bayes_stats.q_ci(:,:,2) - bayes_stats.q_ci(:,:,1);
        fprintf('Bayes: %.2fs\n', toc(t0));
        
        % --- Save Images (using provided helpers) ---
        base_name = sprintf('%s_%sB1', scan.name, mode);
        img_dir = fullfile(out_dir, 'images');
        
        % LRT Maps
        save_t2_img(lrt_maps.q, fullfile(img_dir, [base_name '_LRT_T2.png']), ...
            [strrep(base_name, '_', ' ') ' LRT T2'], t2_disp_range, display_mask);
        
        save_uq_img(lrt_uq, fullfile(img_dir, [base_name '_LRT_Unc.png']), ...
            [strrep(base_name, '_', ' ') ' LRT Unc'], uq_disp_range, display_mask);
            
        % Bayesian Maps
        save_t2_img(bayes_maps.q, fullfile(img_dir, [base_name '_Bayes_T2.png']), ...
            [strrep(base_name, '_', ' ') ' Bayes T2'], t2_disp_range, display_mask);
        
        save_uq_img(bayes_uq, fullfile(img_dir, [base_name '_Bayes_Unc.png']), ...
            [strrep(base_name, '_', ' ') ' Bayes Unc'], uq_disp_range, display_mask);
         
        % --- Extract ROI Statistics ---
        roi_res = extract_roi_stats(roi_masks, sese_t2map, lrt_maps, lrt_stats, bayes_maps, bayes_stats);
        
        % Store for final plotting
        plot_db(s_idx, b_idx).scan = scan.name;
        plot_db(s_idx, b_idx).mode = mode;
        plot_db(s_idx, b_idx).data = roi_res;
    end
end

%% 5. Generate Correlation Plots
fprintf('\nGenerating Correlation Plots...\n');
fs = 14; lw = 1.5;

% Create a figure for each B1 Mode
for b_idx = 1:length(b1_modes)
    mode = b1_modes{b_idx};
    fig = figure('Name', sprintf('T2 Correlation (%s B1)', mode), 'Color', 'w', ...
                 'Position', [100, 100, 1000, 500]);
    
    % Subplots for LRT and Bayesian
    methods = {'LRT', 'Bayesian'};
    
    for m = 1:2
        subplot(1, 2, m); hold on; grid on; box on;
        method = methods{m};
        
        % Identity Line
        plot([0 150], [0 150], 'k--', 'LineWidth', 1, 'HandleVisibility', 'off');
        
        % Scan Markers
        markers = {'o', 's', '^'}; 
        colors  = lines(length(scans));
        
        for s_idx = 1:length(scans)
            if isempty(plot_db(s_idx, b_idx).scan), continue; end % Skip if scan failed
            
            res = plot_db(s_idx, b_idx).data;
            ref_t2 = res.ref_mean;
            
            if strcmp(method, 'LRT')
                est_t2 = res.lrt_mean;
                ci_low = res.lrt_ci_low;
                ci_high = res.lrt_ci_high;
            else
                est_t2 = res.bayes_mean;
                ci_low = res.bayes_ci_low;
                ci_high = res.bayes_ci_high;
            end
            
            % Error Bar Plot (Vertical errors only)
            err_neg = est_t2 - ci_low;
            err_pos = ci_high - est_t2;
            
            errorbar(ref_t2, est_t2, err_neg, err_pos, ...
                markers{s_idx}, 'Color', colors(s_idx,:), ...
                'MarkerFaceColor', colors(s_idx,:), 'LineWidth', lw, ...
                'CapSize', 8, 'DisplayName', strrep(scans(s_idx).name, '_', ' '));
        end
        
        xlabel('Reference SESE T_2 (ms)', 'FontSize', fs);
        ylabel(sprintf('%s Estimated T_2 (ms)', method), 'FontSize', fs);
        title(sprintf('%s (B1: %s)', method, mode), 'FontSize', fs);
        xlim([0 150]); ylim([0 150]);
        axis square; % Force square aspect ratio
        if m == 1, legend('Location', 'northwest', 'FontSize', 10); end
    end
    
    saveas(fig, fullfile(out_dir, 'plots', sprintf('Correlation_Plot_%sB1.png', mode)));
end

fprintf('Done. Results saved to %s\n', out_dir);

File: python_src/run_phantom_t2.py
"""
T2 Phantom Validation Script

Loads T2-weighted phantom data and compares mapping methods against a 
Single-Echo Spin-Echo (SESE) reference. Evaluates performance under:
- Free B1 fitting
- Constrained B1 fitting (using external maps)
"""
import os
import time
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

from utils import load_mat, estimate_noise_covariance, crop_image, extract_roi_stats, save_map
from solvers import fit_mri_params_lrt, fit_mri_params_bayesian

# --- Configuration ---
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.dirname(SCRIPT_DIR)
DATA_DIR = os.path.join(REPO_ROOT, 'data', 'phantom_t2')
DICT_DIR = os.path.join(REPO_ROOT, 'data', 'dictionaries')

USE_IDENTITY_COV = False
FOV = (96, 96)
ALPHA = 0.05
TRUNC_FACTOR = 3.0
RNG_T2 = (0, 150)
RNG_UQ = (0, 40)

OUT_DIR = os.path.join(REPO_ROOT, 'python_output', 'phantom_t2_results_identity' if USE_IDENTITY_COV else 'phantom_t2_results')
os.makedirs(os.path.join(OUT_DIR, 'images'), exist_ok=True)
os.makedirs(os.path.join(OUT_DIR, 'plots'), exist_ok=True)

scans = [
    {'name': '8192_NUFFT', 'file': 'scan_8192_nufft.mat', 'dict': 't2_phantom_TE.mat'},
    {'name': '384_LLR',    'file': 'scan_384_llr.mat',    'dict': 't2_phantom_PC.mat'},
    {'name': '192_LLR',    'file': 'scan_192_llr.mat',    'dict': 't2_phantom_PC.mat'}
]
b1_modes = ['Free', 'Range']

def main():
    print("Loading Reference Data...")
    try:
        ref_data = load_mat(os.path.join(DATA_DIR, 'reference_maps.mat'))
        roi_data = load_mat(os.path.join(DATA_DIR, 'roi_masks.mat'))
        
        sese_t2 = crop_image(ref_data['sese_t2map'], FOV)
        sese_con = crop_image(ref_data['sese_contrast'], FOV)
        tfl_b1 = crop_image(ref_data['tfl_b1map'], FOV)
        
        disp_mask = sese_con[:, :, 0] > 0.1 * np.max(sese_con)
    except Exception as e:
        print(f"Error loading references: {e}")
        return

    plot_db = {} 

    for scan in scans:
        print(f"\n--- Processing Scan: {scan['name']} ---")
        try:
            D = load_mat(os.path.join(DICT_DIR, scan['dict']))['D']
            mat_file = load_mat(os.path.join(DATA_DIR, scan['file']))
            contrast = mat_file['contrast']
            
            # --- Strict Header Check ---
            if 'header' not in mat_file:
                raise ValueError(f"File {scan['file']} does not contain 'header'. Cannot determine Echo Spacing.")
            
            header = mat_file['header']
            if 'esp' not in header or 'etl' not in header:
                raise ValueError(f"Header in {scan['file']} missing 'esp' or 'etl'.")
                
            esp = float(header['esp'])
            etl = int(header['etl'])
            
            # Construct TE Array based on File Header
            te_array = np.arange(1, etl + 1) * esp
            
            contrast = crop_image(contrast.astype(np.complex128), FOV)
            
            if USE_IDENTITY_COV:
                sigma = np.eye(contrast.shape[2])
            else:
                sigma = estimate_noise_covariance(contrast, frame_size=10)
                
            for mode in b1_modes:
                print(f"   > Mode: B1 {mode}")
                
                ops = {
                    'alpha': ALPHA,
                    'te_truncation': (contrast.shape[2] == etl),
                    'te_array': te_array,
                    'trunc_factor': TRUNC_FACTOR
                }
                
                if mode == 'Range':
                    ops['b1_mode'] = 'range'
                    # Constrain B1 search to +/- 10% of the measured map
                    b1_stack = np.stack([0.9 * tfl_b1, 1.1 * tfl_b1], axis=-1)
                    ops['b1_input'] = b1_stack
                else:
                    ops['b1_mode'] = 'none'
                
                # --- Solvers ---
                t0 = time.time()
                _, lrt_res = fit_mri_params_lrt(contrast, sigma, D, ops)
                print(f"     LRT: {time.time()-t0:.2f}s")
                
                t0 = time.time()
                _, bayes_res = fit_mri_params_bayesian(contrast, sigma, D, ops)
                print(f"     Bayes: {time.time()-t0:.2f}s")
                
                # --- Reshape Results (order='F') ---
                nx, ny = FOV
                
                lrt_t2 = lrt_res['q'].reshape(nx, ny, order='F')
                bayes_t2 = bayes_res['q'].reshape(nx, ny, order='F')
                
                lrt_ci = lrt_res['q_ci'].reshape(nx, ny, 2, order='F')
                bayes_ci = bayes_res['q_ci'].reshape(nx, ny, 2, order='F')
                
                lrt_uq = lrt_ci[:,:,1] - lrt_ci[:,:,0]
                bayes_uq = bayes_ci[:,:,1] - bayes_ci[:,:,0]

                # --- Save Images ---
                base = f"{scan['name']}_{mode}B1"
                save_map(lrt_t2, os.path.join(OUT_DIR, 'images', f"{base}_LRT_T2.png"), 
                         f"{base.replace('_',' ')} LRT T2", RNG_T2, disp_mask, 'T2')
                save_map(lrt_uq, os.path.join(OUT_DIR, 'images', f"{base}_LRT_Unc.png"), 
                         f"{base.replace('_',' ')} LRT Unc", RNG_UQ, disp_mask, 'UQ')
                save_map(bayes_t2, os.path.join(OUT_DIR, 'images', f"{base}_Bayes_T2.png"), 
                         f"{base.replace('_',' ')} Bayes T2", RNG_T2, disp_mask, 'T2')
                save_map(bayes_uq, os.path.join(OUT_DIR, 'images', f"{base}_Bayes_Unc.png"), 
                         f"{base.replace('_',' ')} Bayes Unc", RNG_UQ, disp_mask, 'UQ')
                         
                # --- Stats ---
                lrt_struct = {'q': lrt_t2, 'q_ci': lrt_ci}
                bayes_struct = {'q': bayes_t2, 'q_ci': bayes_ci}
                
                stats = extract_roi_stats(roi_data['roi_masks'], sese_t2, lrt_struct, bayes_struct)
                plot_db[(scan['name'], mode)] = stats
                
        except Exception as e:
            print(f"Skipping {scan['name']} due to error: {e}")

    generate_plots(plot_db, scans, b1_modes)

def generate_plots(db, scans, modes):
    print("\nGenerating Correlation Plots...")
    
    colors = ['tab:blue', 'tab:orange', 'tab:green']
    markers = ['o', 's', '^']

    for mode in modes:
        plt.figure(figsize=(10, 5))
        
        for m_idx, method_key in enumerate(['LRT', 'Bayesian']):
            plt.subplot(1, 2, m_idx+1)
            plt.plot([0, 150], [0, 150], 'k--', alpha=0.5)
            
            for s_idx, scan in enumerate(scans):
                key = (scan['name'], mode)
                if key not in db: continue
                
                res = db[key]
                x = res['ref_mean']
                
                if method_key == 'LRT':
                    y = res['lrt_mean']
                    ci_lo = res['lrt_ci_low']
                    ci_hi = res['lrt_ci_high']
                else:
                    y = res['bayes_mean']
                    ci_lo = res['bayes_ci_low']
                    ci_hi = res['bayes_ci_high']
                
                plt.vlines(x, ci_lo, ci_hi, colors=colors[s_idx], alpha=0.5, linewidth=1.5)
                plt.scatter(x, y, color=colors[s_idx], marker=markers[s_idx], 
                            label=scan['name'].replace('_', ' '), zorder=5)
            
            plt.title(f"{method_key} (B1: {mode})")
            plt.xlabel("Reference SESE T2 (ms)")
            plt.ylabel(f"Estimated T2 (ms)")
            plt.xlim(0, 150); plt.ylim(0, 150)
            plt.gca().set_aspect('equal')
            plt.grid(True, alpha=0.3)
            if m_idx == 0: plt.legend()
            
        plt.tight_layout()
        plt.savefig(os.path.join(OUT_DIR, 'plots', f'Correlation_Plot_{mode}B1.png'))
    
    print(f"Done. Output saved to {OUT_DIR}")

if __name__ == "__main__":
    main()
